### PR TITLE
Add MidMeasureMP Kokkos

### PR DESCRIPTION
**Pull request by tomlqc**
### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [ ] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [ ] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
Same as https://github.com/PennyLaneAI/pennylane-lightning/pull/621 for Lightning-Kokkos

**Description of the Change:**
- [ ] Add method to obtain the probability in the z-basis on a given wire.
- [ ] Add RNG attribute.
- [ ] Add method to init the RNG attribute with a given seed.
- [ ] Add method to sample from a binomial distribution.
- [ ] Add method to collapse the state vector onto a given branch.
- [ ] Add method which implements the MidMeasureMP workflow.
- [ ] Add bindings to the MidMeasureMP method.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

**Related Shortcut Story:**
[sc-58334]

### CODE REVIEW

File: pennylane_lightning/core/src/utils/TestHelpers.hpp [hunk 1]
```diff
@@ -288,6 +288,19 @@ auto createZeroState(size_t num_qubits) -> TestVector<ComplexT> {
     return res;
 }
 
+/**
+ * @brief create |+>^N
+ */
+template <typename ComplexT>
+auto createPlusState_(size_t num_qubits) -> TestVector<ComplexT> {
```
Comment (tomlqc) [inline pennylane_lightning/core/src/utils/TestHelpers.hpp:295]: I added this modified version of `createPlusState()` as I needed a `ComplexT` based template for the Kokkos unittest.
Should I try to replace the `PrecisionT` version?

**Comment by tomlqc** [comment]
@vincentmr you may have a look on the C++ implementation, especially about the functor for Kokkos.

File: pennylane_lightning/core/src/simulators/lightning_kokkos/gates/GateFunctorsGenerator.hpp [hunk 2]
```diff
@@ -972,4 +972,35 @@ template <class PrecisionT, bool adj = false> struct generatorMultiRZFunctor {
             1 - 2 * int(Kokkos::Impl::bit_count(k & wires_parity) % 2));
     }
 };
+
+template <class PrecisionT> struct collapseFunctor {
```
Comment (vincentmr) [inline pennylane_lightning/core/src/simulators/lightning_kokkos/gates/GateFunctorsGenerator.hpp:976]: Generators are gate derivatives (up to a factor). This should rather go in the gate functor module. As a rule of thumb, if the constructor does not do any preprocessing a KOKKOS_LAMBDA should be used instead. I would therefore move the implementation to the SV module instead of the gate module.

File: pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp [hunk 3]
```diff
@@ -757,6 +757,60 @@ class StateVectorKokkos final
...
+    void normalize() {
+        KokkosVector sv_view =
+            getView(); // circumvent error capturing this with KOKKOS_LAMBDA
+
+        // TODO: @tomlqc what about squaredNorm()
+        PrecisionT squaredNorm = 0.0;
+        Kokkos::parallel_reduce(
+            sv_view.size(),
+            KOKKOS_LAMBDA(const size_t i, PrecisionT &sum) {
+                sum += std::norm<PrecisionT>(sv_view(i));
```
Comment (vincentmr) [inline pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp:800]: `std::norm` is a host function, so it won't work on GPUs for example. We should use the Kokkos types' methods within kernels.
```suggestion
                sum += norm(sv_view(i));
```

File: pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp [hunk 4]
```diff
@@ -757,6 +757,60 @@ class StateVectorKokkos final
...
+            },
+            squaredNorm);
+
+        PL_ABORT_IF(squaredNorm <
+                        std::numeric_limits<PrecisionT>::epsilon() * 1e2,
+                    "vector has norm close to zero and can't be normalized");
+
+        std::complex<PrecisionT> inv_norm = 1. / std::sqrt(squaredNorm);
```
Comment (vincentmr) [inline pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp:808]: Let's use `Kokkos::complex` (or simply `ComplexT`) and Kokkos functions (like `sqrt`), although in this case this is likely OK since everything should be implicitly casted to `Kokkos::complex` inside the kernel.

File: .github/workflows/tests_gpu_cuda.yml [hunk 5]
```diff
@@ -315,7 +315,7 @@ jobs:
           OMP_PROC_BIND: false
         run: |
           cd main/
-          PL_DEVICE=lightning.qubit python -m pytest tests/ $COVERAGE_FLAGS
+          PL_DEVICE=lightning.qubit python -m pytest tests/ -k "not test_native_mcm" $COVERAGE_FLAGS
```
Comment (vincentmr) [inline .github/workflows/tests_gpu_cuda.yml:318]: `test_native_mcm` is expensive, we only need to run it once in `tests_linux.yml`.

File: .github/workflows/tests_gpu_kokkos.yml [hunk 6]
```diff
@@ -325,7 +325,7 @@ jobs:
           OMP_PROC_BIND: false
         run: |
           cd main/
-          PL_DEVICE=lightning.qubit python -m pytest tests/ $COVERAGE_FLAGS
+          PL_DEVICE=lightning.qubit python -m pytest tests/ -k "not test_native_mcm" $COVERAGE_FLAGS
```
Comment (vincentmr) [inline .github/workflows/tests_gpu_kokkos.yml:328]: Same.

File: .github/workflows/tests_linux.yml [hunk 7]
```diff
@@ -353,8 +354,8 @@ jobs:
           python -m pip install pytest-xdist
           cd main/
           DEVICENAME=`echo ${{ matrix.pl_backend }} | sed "s/_/./g"`
-          OMP_NUM_THREADS=1 PL_DEVICE=${DEVICENAME} python -m pytest -n auto tests/ -k "not unitary_correct" $COVERAGE_FLAGS
-          PL_DEVICE=${DEVICENAME} python -m pytest tests/ -k "unitary_correct" $COVERAGE_FLAGS --cov-append
+          OMP_NUM_THREADS=1 PL_DEVICE=${DEVICENAME} python -m pytest -n auto tests/ -k "not unitary_correct and not test_native_mcm" $COVERAGE_FLAGS
```
Comment (vincentmr) [inline .github/workflows/tests_linux.yml:357]: Etc.

File: pennylane_lightning/lightning_kokkos/lightning_kokkos.py [hunk 8]
```diff
@@ -136,6 +138,8 @@ def _kokkos_configuration():
         "QFT",
         "ECR",
         "BlockEncode",
+        "MidMeasureMP",
+        "Conditional",
```
Comment (maliasadi) [inline pennylane_lightning/lightning_kokkos/lightning_kokkos.py:142]: These are not regular PL operations and this will break Catalyst. We should remove them from here and use class-instance comparison in the decomposition preprocessing step to preserve them in the tape. Please check my fix for LQ in PR #665 to see how these ops were removed from the list. 
```suggestion
```

File: pennylane_lightning/lightning_kokkos/lightning_kokkos.py
Comment (maliasadi) [inline pennylane_lightning/lightning_kokkos/lightning_kokkos.py:1]: There are still several lines not tested by tests. Please consider adding more tests to cover the edge cases of the support.

**Review by maliasadi — COMMENTED** [review]
Nice work!

File: pennylane_lightning/core/src/simulators/lightning_kokkos/tests/Test_StateVectorLKokkos.cpp [hunk 9]
```diff
@@ -308,4 +308,72 @@ TEMPLATE_TEST_CASE("StateVectorKokkos::StateVectorKokkos",
...
+                              expected_state[branch][wire].size(), eps));
+    }
+}
+
+TEMPLATE_TEST_CASE("StateVectorKokkos::normalize", "[StateVectorKokkos]", float,
+                   double) {
+    using PrecisionT = TestType;
+    using ComplexT = typename StateVectorKokkos<PrecisionT>::ComplexT;
+
+    // TODO @tomlqc use same template for testing all Lightning flavours?
```
[the remark this answers is not in the document]
Comment (vincentmr) [reply pennylane_lightning/core/src/simulators/lightning_kokkos/tests/Test_StateVectorLKokkos.cpp:357]: Let's do it when the L-GPU support gets in for the sake of time.

File: pennylane_lightning/lightning_kokkos/lightning_kokkos.py
Comment (vincentmr) [reply pennylane_lightning/lightning_kokkos/lightning_kokkos.py:1]: This repo has several false positives and some errors we do not test because they are hard to reach. Can you point to specific lines that you think should be tested?

Comment (maliasadi) [reply pennylane_lightning/lightning_kokkos/lightning_kokkos.py:1]: Perhaps testing the `capabilities` and `generate_samples` methods. But looking closely into the warnings, none of them are really about the corner cases :+1:

**Review by maliasadi — APPROVED** [review]
Thanks fellas! :100:

**Review by AmintorDusko — APPROVED** [review]
Nice job!

File: pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp [hunk 10]
```diff
@@ -757,6 +757,54 @@ class StateVectorKokkos final
...
+        normalize();
+    }
+
+    /**
+     * @brief Normalize vector (to have norm 1).
+     */
+    void normalize() {
+        auto sv_view = getView();
+
+        // TODO: @tomlqc what about squaredNorm()
```
[the remark this answers is not in the document]
Comment (vincentmr) [reply pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp]: I'll remove as I think this is obsolete. A couple other TODOs might be good to keep.

**Comment by vincentmr** [comment]
> Nice job!

Thanks for the suggestions @AmintorDusko .
